feat: improve generated comments

willbicks opened the pull request:
This feature improves the comments generated for entities by improving the impact of the `.Comment("")` method on fields and edges. The details and justification are described in #2580.

The following changes are implemented:

- If a comment is specified on a field schema with the `.Comment("")` method, that comment now _replaces_ the default comment instead of being appended to it.
- If a comment is specified on an edge schema with the `.Comment("")` method, that comment replaces the default comment instead of being ignored.
- Documentation of the `.Comment("")` method is added to the getting started guide, and the field and edge schema docs.


Closes #2580

willbicks commented:
@a8m just wanted to flag this for your review in case you hadn't seen it, would love to know if you have any input!

a8m commented:
Thanks for the contribution @willbicks 🙏

I'll review this later today. The CI is red - you need to run `go generate ./...` from the root of the project.

willbicks commented:
@a8m Thanks for the tip, I updated the relevant integration schemas to reflect this change and regenerated all the generated files.

a8m commented on doc/md/getting-started.md:
```diff
@@ -84,9 +84,11 @@ import (
 		field.String("name").
-			Default("unknown"),
+			Default("unknown").
+			Comment("Name (first last) of the user. Defaults to \"unkown\"."),
```
You can add it to the fields/edges pages, no need for this in the getting-started page.

a8m commented on doc/md/getting-started.md:
```diff
@@ -310,7 +312,8 @@ Let's add the `"cars"` edge to the `User` schema, and run `go generate ./ent`:
 	return []ent.Edge{
-		edge.To("cars", Car.Type),
+		edge.To("cars", Car.Type).
+			Comment("Cars owned by this user."),
```
gitto

a8m commented on doc/md/schema-edges.mdx:
```diff
@@ -1416,6 +1416,23 @@ However, you should note, that this is currently an SQL-only feature.
+func (User) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.To("pets", Pet.Type).
+			Comment("Pets that this user is responsible for taking care of. \n 
```
```suggestion
			Comment("Pets that this user is responsible for taking care of.\n 
```

a8m commented on doc/md/schema-fields.md:
```diff
@@ -508,6 +508,23 @@ func (User) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("name").
+			Default("John Doe").
+			Comment("Name of the user. \n If not speicifed, defaults to \"John Doe\"."),
```
```suggestion
			Comment("Name of the user.\n If not specified, defaults to \"John Doe\"."),
```

a8m commented on entc/gen/template/ent.tmpl:
```diff
@@ -207,11 +207,23 @@ func ({{ $receiver }} {{ $slice }}) config(cfg config) {
 {{/* A template for setting the field comment. */}}
 {{- define "model/fieldcomment" }}
-	// {{ $.StructField }} holds the value of the "{{ $.Name }}" field.
 	{{- if $.Comment }}
 		{{- range $line := split $.Comment "\n" }}
 			// {{ $line }}
 		{{- end }}
+    {{- else }}
+        // {{ $.StructField }} holds the value of the "{{ $.Name }}" field.
+	{{- end }}
```
```suggestion
	{{- if $.Comment }}
		{{- range $line := split $.Comment "\n" }}
			// {{ $line }}
		{{- end }}
	{{- else }}
		// {{ $.StructField }} holds the value of the "{{ $.Name }}" field.
	{{- end }}
```

a8m commented on entc/gen/template/ent.tmpl:
```diff
@@ -207,11 +207,23 @@ func ({{ $receiver }} {{ $slice }}) config(cfg config) {
+		{{- range $line := split $.Comment "\n" }}
+			// {{ $line }}
+		{{- end }}
+    {{- else }}
+        // {{ $.StructField }} holds the value of the {{ $.Name }} edge.
 	{{- end }}
```
```suggestion
	{{- else }}
		// {{ $.StructField }} holds the value of the {{ $.Name }} edge.
	{{- end }}
```

a8m commented on entc/integration/edgeschema/ent/schema/tweet.go:
```diff
@@ -31,6 +31,6 @@ func (Tweet) Edges() []ent.Edge {
 			Ref("tweets").
 			Through("tweet_user", UserTweet.Type).
-			Comment("The uniqueness is enforced on the edge schema"),
+			Comment("User who authored this tweet. \nThe uniqueness is enforced on the edge schema"),
```
No need to change this comment, please. The second sentence is more important for this example.

a8m reviewed (approved):
Amazing job!

I added a few small comments. Please address them and I'll merge it. Thanks <3

willbicks commented:
@a8m, I was able to push a new version that addressed your suggestions.

I have a few more places where I think the comment generation functionality could be improved, so I'll be looking into those and submit some requests when I have something.

Thanks for all your work in maintaining this project!

a8m reviewed (approved):
Excellent job 🚀
